src/ui/utils.ts

```diff
@@ -1,7 +1,134 @@
-import {isMobile} from '../utils/platform';
+import {isMobile, isFirefox} from '../utils/platform';
+import UserStorage from '../background/user-storage';
 
 declare const __CHROMIUM_MV3__: boolean;
 
+declare const browser: {
+    theme: {
+        getCurrent: (() => Promise<void>);
+    };
+};
+
+export async function getTheme() {
+    // Check if theme setting should be used
+    await UserStorage.loadSettings();
+    if (UserStorage.settings.themeWithFirefox && isFirefox && typeof browser !== 'undefined' && browser.theme && browser.theme.getCurrent) {
+        setStyle(await browser.theme.getCurrent());
+    }
+}
+
+function parseColor(color: any) {
+    // Create a div, set a color, get it back in unified rgb format
+    var div = document.createElement('div'), m;
+    document.body.appendChild(div);
+    div.style.color = color;
+    m = getComputedStyle(div).color.match(/^rgb\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*\)$/i);
+    div.remove();
+    if(m) return [Number(m[1]),Number(m[2]),Number(m[3])];
+    return [0,0,0];
+}
+
+function setStyle(Theme: any) {
+    if (Theme.colors) {
+        // Get color or use fallback white
+        var Color = Theme.colors.popup_text;
+        if (Color.toString() == "hsl(0, 0%, 100%)") {
+            Color = "rgb(255, 255, 255)";
+        }
+        if (Color.toString() == "hsl(0, 0%, 0%)") {
+            Color = "rgb(0, 0, 0)";
+        }
+
+        // Get 2 different strength highlights
+        const BackgroundColor = "rgba" + Color.slice(3, Color.length - 1) + ", .17)"
+        const ActiveBackgroundColor = "rgba" + Color.slice(3, Color.length - 1) + ", .34)"
+
+        // Check whether theme is light or dark
+        var CRGB = parseColor(Color);
+        var BRGB = parseColor(Theme.colors.popup);
+        var Light = 1;
+        if(BRGB[0] + BRGB[1] + BRGB[2] <= CRGB[0] + CRGB[1] + CRGB[2]) {
+            Light = 0;
+        }
+
+        // Create and apply css
+        var Parser = new DOMParser();
+        var Style = Parser.parseFromString(`
```

Comment:
Parsing HTML strings is open to XSS attacks or other unwanted content. There's `new CSSStyleSheet()` and `adoptedStyleSheets` for creating CSS in JS, if for some reason CSS files cannot be used. Or `document.createElement('style')` and `element.textContent = '...'` could be used as an alternative.